components/ContractViewer/ContractTreeNode.tsx

```diff
@@ -0,0 +1,661 @@
+      updateFuncData({ outputs: [] })
+      return
+    }
+
+    ethCall(data)
+      .then((decoded: any) => {
+        setStatus('')
+        updateFuncData({ outputs: decoded })
+      })
+      .catch((err: any) => {
+        setStatus(err.toString())
+        updateFuncData({ outputs: [] })
+      })
+  }
+
+  const setEncodeStatus = () => {
+    const [data, error] = encodeCalldata()
+    setStatus('calldata: ' + (data || error))
+  }
+
+  return (
+    <TreeItemBasic
+      nodeId={`function_${id}_${funcAbi.name}`}
+      title={funcName}
+      subtitle={subtitle}
+    >
+      <div className="flex flex-col gap-2 text-black-500 my-2 mr-4">
+        <span className="text-xs dark:text-gray-200">
+          4byte selector: {toFunctionSelector(funcAbi)}
+        </span>
+        {funcAbi && <ParamsBox abi={funcAbi} reducer={reducer} />}
+        {(funcAbi.inputs.length > 0 || funcAbi.stateMutability == 'payable') && (
+          <div className="flex gap-1">
+            <Button onClick={setCallStatus} variant="contained">
+              {funcAbi?.outputs?.length > 0
+                ? 'Call'
+                : 'Call (no ret)'}
+            </Button>
+            <Button onClick={setEncodeStatus} variant="contained">
+              Encode
+            </Button>
+            <Button onClick={ethSendTransaction} variant="contained" color="secondary">
+              Send
+            </Button>
+          </div>
+        )}
+        {(funcAbi && funcAbi?.outputs?.length > 0) && <ReturnDataBox abi={funcAbi} reducer={reducer} />}
+        {status && (
+          <input
+            type="button"
+            className="cursor-pointer hover:bg-blue-100 dark:hover:bg-blue-900 text-left whitespace-pre-line break-all text-sm dark:text-gray-100"
+            onClick={() =>
+              navigator.clipboard.writeText(status)
+            }
+            value={'📋 copy'}
+          />
+        )}
+        <p className="text-xs text-gray-500 break-words">
+          {spaceBetween(status || '')}
+        </p>
+      </div>
+    </TreeItemBasic>
+  )
+}
+
+type StorageLayoutItemProps = {
+  id: string
+  address: string
+  storage: {
+    astId: number
+    label: string
+    contract: string
+    offset: number
+    slot: any // string | number
+    type: string
+  }
+  types: {
+    [type: string]: {
+      key: string
+      encoding: string
+      label: string
+      numberOfBytes: string
+      value: string
+    }
+   }
+}
+
+export const StorageLayoutItem = ({ id, address, storage, types }: StorageLayoutItemProps) => {
+  const [status, setStatus] = useState('0x...')
+  const [inputs, setInputs] = useState([])
+
+  const type = types[storage.type]
+
+  // if storage item is a mapping, keyTypes is an array of types of all mapping keys
+  // eg. keys of mapping(address => mapping(uint256 => bytes)) would be ['address', 'uint256']
+  const keyTypes = [...(storage.type.matchAll(/\bt_mapping\((?<key>.+?),/g) || [])].map((m) => m.groups.key)
```

Comment:
I guess using the regex should work just fine, but you could also check if the encoding is a mapping and then read of the key value there.